src/scheduler/Executor.cpp

```diff
@@ -79,99 +119,116 @@ void Executor::executeTasks(std::vector<int> msgIdxs,
-                      }
-                  }));
-            }
+        if (threadPoolThreads.at(threadPoolIdx) == nullptr) {
+            threadPoolThreads.at(threadPoolIdx) = std::make_shared<std::thread>(
+              &Executor::threadPoolThread, this, threadPoolIdx);
+        }
+    }
+}
+
+void Executor::threadPoolThread(int threadPoolIdx)
+{
+    auto logger = faabric::util::getLogger();
+    logger->debug("Thread pool thread {}:{} starting up", id, threadPoolIdx);
+
+    auto& sch = faabric::scheduler::getScheduler();
+    auto& conf = faabric::util::getSystemConfig();
+
+    for (;;) {
+        logger->trace("Thread starting loop {}:{}", id, threadPoolIdx);
+        std::pair<int, std::shared_ptr<faabric::BatchExecuteRequest>> task;
+
+        try {
+            task = threadQueues[threadPoolIdx].dequeue(conf.boundTimeout);
+        } catch (faabric::util::QueueTimeoutException& ex) {
+            // If the thread has had no messages, it needs to
+            // remove itself
+            shutdownThreadPoolThread(threadPoolIdx);
+            break;
+        }
+
+        int msgIdx = task.first;
+
+        // If the thread is being killed, the executor itself
+        // will handle the clean-up
+        if (msgIdx == -1) {
+            logger->debug(
+              "Killing thread pool thread {}:{}", id, threadPoolIdx);
+            break;
+        }
+
+        std::shared_ptr<faabric::BatchExecuteRequest> req = task.second;
+        assert(req->messages_size() >= msgIdx + 1);
+        faabric::Message& msg = req->mutable_messages()->at(msgIdx);
+
+        logger->trace("Thread {}:{} executing task {} ({})",
+                      id,
+                      threadPoolIdx,
+                      msgIdx,
+                      msg.id());
+
+        int32_t returnValue;
+        try {
+            returnValue = executeTask(threadPoolIdx, msgIdx, req);
+        } catch (const std::exception& ex) {
+            returnValue = 1;
+
+            msg.set_outputdata(fmt::format(
+              "Task {} threw exception. What: {}", msg.id(), ex.what()));
+        }
+
+        // Set the return value
+        msg.set_returnvalue(returnValue);
+
+        // Decrement the task count
+        int oldTaskCount = executingTaskCount.fetch_sub(1);
+        assert(oldTaskCount >= 0);
+
+        logger->trace("Task {} finished by thread {}:{} ({} left)",
+                      msg.id(),
+                      id,
+                      threadPoolIdx,
+                      executingTaskCount);
+
+        // Set function result
+        if (req->type() == faabric::BatchExecuteRequest::THREADS) {
+            sch.setThreadResult(msg, returnValue);
+        } else {
+            sch.setFunctionResult(msg);
+        }
+
+        // Notify the scheduler, note that we have to release the claim _after_
+        // resetting, once the executor is ready to be reused
+        if (oldTaskCount == 1) {
+            reset(msg);
+            releaseClaim();
```

Comment:
This is the only slightly modified part, where we have to release the claim to this executor.